### DISCUSSION TITLE
Is there an equivalent to BaseModel.dict(exclude_unset=True) for dataclasses?

### DISCUSSION THREAD

**Discussion by jeffsawatzky** [pydantic/pydantic#5716] [Question] [answered]
I'm using Pydantic dataclasses for my data models. I would like to convert the data instance to a dict, but only include the fields that we explicitly set (even if they are the same as the default value).

Basically, I am trying to achieve this:
https://fastapi.tiangolo.com/tutorial/body-updates/#using-pydantics-exclude_unset-parameter

But with pydantic dataclasses instead of using the BaseModel. Is this possible?

**Comment by samuelcolvin** [comment, marked as answer]
Fundamentally not possible I'm afraid, dataclasses don't store which fields have been set.

**Reply by dmontagu** [reply]
The closest you'll get is probably to use a sentinel object as the default value (you can create a class called `Undefined` or similar, and use an instance of it as the default value). Then, when you serialize, remove the key anywhere the value satisfies `isinstance(value, Undefined)`.

**Reply by jeffsawatzky** [reply]
Dang. I was hoping that the pydantic dataclass wrapper did some magic to track that. Thanks for answering my question.

**Comment by laundmo** [comment]
i've added the following to a base dataclass all my dataclass models inherit from:
```py
def __setattr__(self, __name: str, __value: Any) -> None:
        if not __name.startswith("_"):
            changed: Set[str] = set()
            changed = getattr(self, "_changed", changed)
            changed.add(__name)
            setattr(self, "_changed", changed)
        super().__setattr__(__name, __value)
```

this way i can later use
```py
RootModel[Model](m).model_dump(mode="json", include=getattr(m, "_changed", set()))
```
to achieve a similar effect.

sadly, abusing the way pydantic does exclude_unset internally, by setting `__pydantic_fields_set__` this way, didn't work.
